src/saturn_engine/utils/error_handling.py

```diff
@@ -0,0 +1,188 @@
+from typing import Optional
+from typing import Type
+from typing import cast
+
+import dataclasses
+import re
+from dataclasses import dataclass
+from functools import lru_cache
+from textwrap import dedent
+from types import TracebackType
+
+from saturn_engine.core.api import TopicItem
+from saturn_engine.core.error import ErrorMessageArgs
+from saturn_engine.core.pipeline import PipelineOutput
+from saturn_engine.core.pipeline import PipelineResults
+from saturn_engine.core.topic import TopicMessage
+from saturn_engine.utils import inspect as extra_inspect
+from saturn_engine.utils.traceback_data import TracebackData
+from saturn_engine.worker.executors.bootstrap import RemoteException
+
+
+@dataclass
+class ExceptionDetails:
+    message: str
+    exception_type: Type[BaseException]
+    module: str
+    lineno: int
+    traceback: TracebackData
+
+
+@dataclass
+class ExceptionFilter:
+    message_pattern: Optional[re.Pattern]
+    exception_type: Type[BaseException]
+    module_pattern: Optional[re.Pattern]
+    lineno: int
+
+
+def process_pipeline_exception(
+    outputs: dict[str, list[TopicItem]],
+    exc_value: BaseException,
+    exc_traceback: TracebackType,
+) -> PipelineResults:
+    # Match the exception and consume it if found, else reraise the exception
+    for channel in outputs.keys():
+        if not is_output_error_handler(channel):
+            continue
+
+        exc_filter = parse_warning_filter(channel, escape=False)
+        exc_details = get_exception_details(exc_value, exc_traceback)
+
+        if not does_error_match(exception=exc_details, exception_filter=exc_filter):
+            continue
+
+        return PipelineResults(
+            outputs=[
+                PipelineOutput(
+                    channel=channel,
+                    message=TopicMessage(
+                        args=dataclasses.asdict(
+                            ErrorMessageArgs(
+                                type=exc_details.exception_type.__name__,
+                                module=exc_details.module,
+                                message=str(exc_value),
+                                traceback=exc_details.traceback,
+                            )
+                        )
+                    ),
+                )
+            ],
+            resources=[],
+        )
+    # Unhandled exception
+    raise exc_value
```

Comment:
Don't raise here, you are losing all contextual information attached to the exception. Instead make the return value `Optional` and reraise in the exception handler (Through a naked `raise`)